sw/device/lib/base/memory_hardened.c

```diff
@@ -0,0 +1,670 @@
+   */
+  uintptr_t mask = -(val * 7 != 0); /* val=0 -> 0, val!=0 => 0xfffffff */
+  return (hardened_bool_t)((mask & (uintptr_t)kHardenedBoolFalse) |
+                           ((~mask) & (uintptr_t)kHardenedBoolTrue));
+}
+
+/**
+ * Define machine word and alignment
+ */
+typedef uintptr_t word_t;
+#define WORD_SIZE sizeof(word_t)
+#define WORD_MASK (WORD_SIZE - 1)
+
+static inline size_t min(size_t a, size_t b) { return (a < b) ? a : b; }
+
+/**
+ * The 32 bit LFSR whose maximum length feedback polynomial is represented
+ * as X^32 + X^22 + X^2 + X^1 + 1 will produce 2^32-1 PN sequence.
+ * This LFSR can be initialized with 0,  but can't be initialized with
+ * 0xFFFFFFFF
+ */
+static uint32_t lfsr32(uint32_t seed) {
+  uint32_t bit = (seed ^ (seed >> 1) ^ (seed >> 21) ^ (seed >> 31)) & 1;
+  return (seed << 1) | (bit ^ 1);
+}
+
+/**
+ * Swap byte order between big-endian and little-endian
+ * @param x	Input word
+ * @return x in different endianess
+ */
+static inline uint32_t bswap32(uint32_t x) { return __builtin_bswap32(x); }
+
+/**
+ * Division by multiplication on mod 2^32 inverse.
+ * Result can be larger than interfer division
+ *
+ * @returns ~( p / (2^32/inv))
+ */
+static inline uint32_t div_by_mul(uint32_t p, uint32_t inv) {
+  return (inv) ? (uint32_t)(((uint64_t)p * (uint64_t)inv) >> 32) : p;
+}
+
+/**
+ * Compute *div = i/n, *rem = i%n using multiplication
+ */
+static void inv_mod(size_t i, uint32_t N, uint32_t N_inverse, uint32_t *div,
+                    uint32_t *rem) {
+  /* Divide by multiplication on modular inverse */
+  uint32_t d = div_by_mul(i, N_inverse);
+
+  /* compute mod i by multiplication and subtraction */
+  uint32_t dN = d * N;
+
+  /* in some cases we get result up to +2 */
+  while (i < dN) {
+    dN -= N;
+    d--;
+  }
+
+  if (div)
+    *div = d;
+  *rem = i - dN;
+}
+
+/**
+ * Branchless (bit) ? t : f for bit 0 or 1
+ */
+static inline void *mask_select(uintptr_t bit, void *f, void *t) {
+  bit -= 1; /* 0 -> 0xFFFFFFFF, 1 -> 0 */
+  return (void *)((bit & (uintptr_t)f) | ((~bit) & (uintptr_t)t));
+}
+
+/* Fault resistant memcpy */
+hardened_bool_t memcpy_fr(void *restrict dest, const void *restrict src,
+                          size_t len) {
+  uint8_t *d = (uint8_t *)dest;
+  const uint8_t *s = (const uint8_t *)src;
+  word_t *dw;
+  const word_t *sw;
+  uint8_t *head;
+  uint8_t *const tail = d + len;
+
+  /* Set 'body' to the last word boundary */
+  word_t *const body = (word_t *)((uintptr_t)tail & ~WORD_MASK);
+
+  /* src and/or dest do not align on word boundary */
+  if (((uintptr_t)dest ^ (uintptr_t)src) & WORD_MASK || (len < WORD_SIZE)) {
+    /* Misaligned. no body, no tail. */
+    head = tail;
+  } else {
+    /* Set 'head' to the first word boundary */
+    head = (uint8_t *)(((uintptr_t)d + WORD_MASK) & ~WORD_MASK);
+  }
+
+  /* Copy misaligned head */
+  while (d < head)
+    *(d++) = *(s++);
+
+  /* Copy aligned body (if any) */
+  dw = (word_t *)d;
```

Comment:
with this one I'm struggling to understand where aliasing issue comes from. read32 and write32 can be used if it really makes sense. I rewrote code to use single pointer of type uintptr_t and accesses as byte or word wherever address is aligned or not (first byte access, then if aligned region is present, use word access, but still it's same pointer/address). It allows to avoid definition of differently typed pointers and ensure no overlap.